Add GPU implementation for tf.segment_sum.

tjingrant opened the pull request:
As per #11228, a GPU version of segment sum has been created. 
## Performance 
On a Tesla K40c GPU, the following performance tests have been performed. 
- The test case sizes are represented using 3-tuples of integers. The first two integers correspond to the outer and inner dimension of the input data and the last integer corresponds to the outer dimension of the output data, which is also the total number of segments. 
- The first three rows of data are obtained by running reductions on float32 input type whereas the last three rows of data are obtained by running reductions on float64 input type. 
- We use the GPU version of unsorted segment reduction as the baseline. 
- During experiments, they share the same input data and outputs are compared for consistency.

|  | (1024, 1024, 128)            |   (2048, 2048, 256)          |    (4096, 4096, 512)         | 
|-------------------------------------------------------------|-------------|-------------|-------------| 
| t_unsorted                                                  | 84.942us    | 332.12us    | 1.3170ms    | 
| t_sorted                                                    | 71.047us    | 264.16us    | 1.0391ms    | 
| speedup                                                     | 1.20  | 1.26 | 1.27  | 
|                                                             |             |             |             | 
| t_unsorted                                                  | 160.69us    | 594.93us    | 2.2895ms    | 
| t_sorted                                                    | 106.94us    | 395.19us    | 1.5662ms    | 
| speedup                                                     | 1.50 | 1.51 | 1.46 | 

## Known Limitations:
- We do not check segment_id to make sure that they are within [0, num_outputs-1].
- We do not verify that segment_id is increasing.

tjingrant committed:
Add GPU implementation for tf.segment_sum.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

tjingrant commented:
I signed it!

zheng-xq commented on tensorflow/core/kernels/segment_reduction_ops.cc:
```diff
@@ -183,6 +186,54 @@ class SegmentReductionOp : public OpKernel {
+    const int64 num_indices = segment_ids.NumElements();
+
+    Index output_rows;
+    context->eigen_device<GPUDevice>().memcpyDeviceToHost(
```
This introduces a device/host sync. TensorFlow tries very hard to avoid that. 

An example on how to address such issue: 

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/where_op.cc#L253

Another common practice is to check the indices on GPU, and do nothing with invalid indicies.

zheng-xq commented on tensorflow/core/kernels/segment_reduction_ops.cc:
```diff
@@ -183,6 +186,54 @@ class SegmentReductionOp : public OpKernel {
+#ifdef GOOGLE_CUDA
+//  SegmentSumGPUOp is a segment sum operator implemented for GPU only.
+template <class T, class Index>
+class SegmentSumGPUOp : public OpKernel {
```
The convention is to overload "SegmentReductionOp" for GPUDevice

tjingrant replied:
Hi, the current implementation of SegmentReductionOps is actually rather specialized for non-GPU devices. I'm not sure if the usual different-functor-for-different-device approach is the best way to go here. I actually started this implementation by porting existing code to be reused on GPU but later gave up because it's very inefficient for GPU device; in addition, the original implementation takes data input as non-primitive data types (i.e., TensorMap) which means if I were to use GPU specialized functors, I may have to introduce some extra data types into the .cu file and risk contaminating the .cu file with unnecessary header files; again, with these considerations, I'm not sure overloading SegmentReductionOp is the best approach. Nevertheless, I'm more than happy to consult with your thoughts on this.

zheng-xq commented on tensorflow/core/kernels/segment_reduction_ops.cc:
```diff
@@ -183,6 +186,54 @@ class SegmentReductionOp : public OpKernel {
+    auto data_ptr = input.template flat<T>().data();
+    auto segment_flat = segment_ids.flat<Index>();
+
+    functor_(context, context->eigen_device<GPUDevice>(), output_rows,
```
Since the functor is trivial, we don't have to cache it.

tjingrant replied:
Thanks! Will update that.

zheng-xq commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -54,6 +54,74 @@ __device__ __forceinline__ void AccumulateInto(
+// rows of input data and all reduction elements share one inner
+// dimension index.
+#define OUTER_DIM_TILE_SIZE 8
+#define CEIL_DIV(x, y) (1 + (((x)-1) / (y)))
```
This produces the wrong answer when x == 0. The typical way is to write (x + y - 1) / y. 

In TF, the convention is to use Eigen::divup

tjingrant replied:
Thanks for picking up this mistake.

zheng-xq commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -80,6 +148,44 @@ __global__ void UnsortedSegmentSumCustomKernel(
+  const Index input_inner_dim_size = input_total_size / input_outer_dim_size;
+
+  const Index input_outer_dim_num_strip =
+      1 + (input_outer_dim_size - 1) / OUTER_DIM_TILE_SIZE;
```
Same here. Use div up

tjingrant replied:
Will do.

zheng-xq reviewed (commented):
First wave of comments. I'll take a closer look at the details later.

Overall, please add functional tests and benchmarks to make sure it is actually faster across a range of different shapes with different aspect ratios.

Also we need to add more data shapes and randomly generated data to make sure there is no hidden race conditions in the logic.

zheng-xq commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -54,6 +54,74 @@ __device__ __forceinline__ void AccumulateInto(
+        if (last_output_segment_id == first_segment_id) {
+          AccumulateInto<T>(output + output_index, sum);
+        } else {
+          *(output + output_index) = sum;
```
I took a closer look at this code. I think we should definitely add more tests to make sure there is no hidden race condition. Also adding a benchmark showing there is no performance regression against unsorted_segment_sum is good. 

Please make sure try a few different tensor shape aspect ratio in the benchmarks.

zheng-xq commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -54,6 +54,74 @@ __device__ __forceinline__ void AccumulateInto(
+    T sum = T(0);
+    Index first_segment_id = segment_ids[input_outer_dim_index_base];
+    Index last_output_segment_id = output_outer_dim_size;
+    const Index actual_stripe_height = MIN(
```
As for "MIN", tf convention is "min"

zheng-xq commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -54,6 +54,74 @@ __device__ __forceinline__ void AccumulateInto(
+                                             const Index* segment_ids,
+                                             const T* input, T* output) {
+  const Index input_outer_dim_num_strip =
+      CEIL_DIV(input_outer_dim_size, OUTER_DIM_TILE_SIZE);
```
Integer division is quite expensive, maybe pass them in.

tjingrant committed:
Reformat comments for stylistic consistency.

tjingrant committed:
Refactor segment sum to compute asynchronously.

tjingrant committed:
Fixed various issues raised in the comments.

tjingrant committed:
Migrate sync assertions to async assertions and fix errors with empty input.

tjingrant commented:
I have updated the pertinent code. Performance tests with more aspect ratios are conducted and recorded in this [gist](https://gist.github.com/tjingrant/3113562d6e9fc8738487ca91847ffd16). The results of both segment_sum and unsorted_segment_sum are always compared for consistency and therefore this also serves as a correctness test for potential race conditions.

In general, the execution time of sorted segment sum cuda kernel is uniformly shorter than that of unsorted segment sum cuda kernel when the problem size is medium or large. It is clear that tiling comes with an inherent cost and therefore when the problem size is small, the performance penalty imposed by tiling will be more noticeable but hopefully nevertheless small. Of all the problem sizes tested (small and large problem sizes alike), the average speed up factor is 1.17 for fp32 and 1.38 for fp64. In general, for fp32 typed segment summation, the break even point is around 512x512 and for any problem sizes beyond this point, using sorted segment sum will be profitable. For fp64, this point is very low and we actually did not see executions of any problem sizes during which sorted segment sum is slower.

It is also worth noting that sorted segment sum requires stream synchronization in order to retrieve output size (a scalar) from device memory. Therefore the overall break even point is somewhat higher but since this cost is constant, it will not impact the execution time of kernels doing more significant work. Please let me know if there is any way to mitigate this problem. Specifically, 

- is it possible to access the input array on the host before it is moved to the device? 
- Or slightly worse, is it possible to have duplicate copies of array at both host and device memory?

rmlarsen commented:
@zheng-xq can you please take another look?

tjingrant committed:
Add GPU tests and change test datatype to accommodate GPU kernel input requirement.

tjingrant committed:
Fix problems with error reporting.

rmlarsen commented:
@tensorflow-jenkins test this please

tjingrant commented:
Hi, judging from the console output, I'm not sure if the test failure is caused by my changes.

`//bazel_pip/tensorflow/contrib/timeseries/python/timeseries:state_management_test FAILED in 13.3s`

This does not look like anything I've touched.

rmlarsen commented:
@tjingrant indeed, that appears unrelated to your change.

rmlarsen commented:
@zheng-xq XQ, this appears to pass (except for known flake). Can you approve it?

zheng-xq commented:
LGTM

@tjingrant, I think it is a good idea to check in your benchmark so the changes are self-sufficient. But I will leave it up to you whether you want to make it into a follow-up CL.

tjingrant committed:
Add benchmarks.

tjingrant commented:
Updated with benchmarks included.

tjingrant commented:
ping @zheng-xq , @rmlarsen

nolanliou commented:
I do a quick test about your code with the [script](https://gist.github.com/nolanliou/a1dc1df3d7d6267a0177b6749ccdf696). the result show there is not as good as unsorted_segment_sum with larger scale of segment_ids.
```
unsorted_segment_sum: 5.616264s
sorted_segment_sum: 7.558630s
unique ids: 6287
```

tjingrant commented:
Hi @nolanliou , thanks for checking out my code! As for your observation, as you probably know, there are a lot of factors that could affect the performance and after all, your ratio of reductions performed to your total data footprint is quite low and therefore in this case I do not expect my code to perform well. I encourage you to explore more problem sizes and reduction to data footprint ratio. Apart from that, the **main reason** you are seeing such a dramatic performance discrepancy is that you are not using the unsorted segment reductions correctly and probably it's not producing the correct results. Try placing it on a CPU device and you will see the error messages.

zheng-xq commented on tensorflow/python/kernel_tests/segment_reduction_ops_test.py:
```diff
@@ -641,6 +646,56 @@ def testGradientSegmentsInvalid4(self):
         with self.assertRaisesOpError(r"Segment id 0 out of range \[0, 0\)"):
           s.eval()
 
+def type_to_str(t):
```
Minor nitpicking: maybe np_type_to_str?

zheng-xq commented on tensorflow/python/kernel_tests/segment_reduction_ops_test.py:
```diff
@@ -641,6 +646,56 @@ def testGradientSegmentsInvalid4(self):
+
+class SegmentReductionOpBenchmark(test.Benchmark):
+
+  def benchmarkSegmentSumGPU(self):
```
Check "is_gpu_available"?

zheng-xq commented on tensorflow/python/kernel_tests/segment_reduction_ops_test.py:
```diff
@@ -641,6 +646,56 @@ def testGradientSegmentsInvalid4(self):
+      seg_ids = np.sort(np.random.randint(
+          output_outer_dim, size=outer_dim))
+
+      op_functors = [lambda vc, vs:
```
Most benchmarks would split them into two different sub-benchmarks. They can share the underlying implementation

zheng-xq commented on tensorflow/python/kernel_tests/segment_reduction_ops_test.py:
```diff
@@ -641,6 +646,56 @@ def testGradientSegmentsInvalid4(self):
+
+  def benchmarkSegmentSumGPU(self):
+    repeat = 10
+    outer_dim_options = [2**x for x in range(7, 14, 2)]
```
Please check large size with less alignments as well.

zheng-xq commented on tensorflow/python/kernel_tests/segment_reduction_ops_test.py:
```diff
@@ -641,6 +646,56 @@ def testGradientSegmentsInvalid4(self):
+            t.append(r["wall_time"])
+
+      # print out the speed up factor for each test
+      print("{:d}\t{:d}\t{:d}\t{:f}".format(
```
We really care about the speedup before and after your CL. Please include the comparison before and after in your commit description. Thanks!

tjingrant replied:
Hi @zheng-xq, there's no (GPU) implementation before my commit. This commit enables segment_sum on GPU. The speedup to the base-line here is another thing. Can you double-check this (the speed up against baseline, not impl before my commit) is what you want?

zheng-xq replied:
Fair point. In that case, the comparison report in this issue is good enough. If you can include in the commit description, that is helpful. The benchmarks should be separate though.

tjingrant committed:
Benchmark results against baseline unsorted impl.

The columns are: datatypes, outer dimension, output outer dimension, inner dimension, execution time difference against baseline impl as a percentage of the time taken by the baseline impl (positive values mean faster execution than baseline).

fp32    512     256     512     -11.8632
fp64    512     256     512     -10.1854
fp32    512     256     2048    -6.2147
fp64    512     256     2048    -0.106
fp32    512     256     8192    -0.0867
fp64    512     256     8192    3.6285
fp32    512     256     1120    -10.9163
fp64    512     256     1120    -1.3509
fp32    512     256     1215    -2.0428
fp64    512     256     1215    -7.884
fp32    512     256     1856    -7.3159
fp64    512     256     1856    -0.1011
fp32    512     256     1302    -3.7802
fp64    512     256     1302    4.6675
fp32    512     256     1329    -13.2275
fp64    512     256     1329    4.3505
fp32    512     256     1531    -7.5993
fp64    512     256     1531    -3.5371
fp32    512     256     1313    -5.0677
fp64    512     256     1313    -1.368
fp32    512     256     1672    -0.0907
fp64    512     256     1672    4.5809
fp32    512     256     1851    -10.2862
fp64    512     256     1851    2.3119
fp32    512     256     1584    -10.3406
fp64    512     256     1584    1.4481
fp32    512     64      512     -6.3544
fp64    512     64      512     -18.4343
fp32    512     64      2048    -9.6639
fp64    512     64      2048    0.0714
fp32    512     64      8192    1.2097
fp64    512     64      8192    10.4839
fp32    512     64      1120    -20.1102
fp64    512     64      1120    -5.0784
fp32    512     64      1215    -6.4061
fp64    512     64      1215    -14.1781
fp32    512     64      1856    2.4221
fp64    512     64      1856    -8.4205
fp32    512     64      1302    -10.2403
fp64    512     64      1302    -7.0577
fp32    512     64      1329    -10.515
fp64    512     64      1329    -4.4899
fp32    512     64      1531    -18.4045
fp64    512     64      1531    9.5982
fp32    512     64      1313    -17.0858
fp64    512     64      1313    -2.02
fp32    512     64      1672    -6.1997
fp64    512     64      1672    1.1616
fp32    512     64      1851    -2.3241
fp64    512     64      1851    -1.0585
fp32    512     64      1584    -7.2199
fp64    512     64      1584    -1.0865
fp32    512     16      512     -14.8754
fp64    512     16      512     -10.987
fp32    512     16      2048    -18.3725
fp64    512     16      2048    1.5949
fp32    512     16      8192    2.163
fp64    512     16      8192    12.5431
fp32    512     16      1120    2.2558
fp64    512     16      1120    -2.7135
fp32    512     16      1215    -12.7228
fp64    512     16      1215    11.0343
fp32    512     16      1856    -6.986
fp64    512     16      1856    7.0687
fp32    512     16      1302    -9.3881
fp64    512     16      1302    -8.2974
fp32    512     16      1329    -11.5103
fp64    512     16      1329    18.9707
fp32    512     16      1531    -14.3721
fp64    512     16      1531    9.6774
fp32    512     16      1313    -16.5546
fp64    512     16      1313    11.7528
fp32    512     16      1672    -10.3689
fp64    512     16      1672    15.1197
fp32    512     16      1851    -11.6021
fp64    512     16      1851    8.2983
fp32    512     16      1584    -13.6702
fp64    512     16      1584    9.4635
fp32    2048    1024    512     -5.6482
fp64    2048    1024    512     1.45
fp32    2048    1024    2048    -0.066
fp64    2048    1024    2048    3.6549
fp32    2048    1024    8192    4.3953
fp64    2048    1024    8192    5.0636
fp32    2048    1024    1120    2.3119
fp64    2048    1024    1120    3.4102
fp32    2048    1024    1215    1.6251
fp64    2048    1024    1215    2.4538
fp32    2048    1024    1856    1.4219
fp64    2048    1024    1856    5.2966
fp32    2048    1024    1302    0.4938
fp64    2048    1024    1302    3.6871
fp32    2048    1024    1329    2.7753
fp64    2048    1024    1329    4.2955
fp32    2048    1024    1531    1.8766
fp64    2048    1024    1531    4.4579
fp32    2048    1024    1313    0.6639
fp64    2048    1024    1313    4.5556
fp32    2048    1024    1672    1.1072
fp64    2048    1024    1672    3.8653
fp32    2048    1024    1851    1.1566
fp64    2048    1024    1851    3.6434
fp32    2048    1024    1584    0.7806
fp64    2048    1024    1584    4.3265
fp32    2048    256     512     -10.7236
fp64    2048    256     512     1.011
fp32    2048    256     2048    2.2321
fp64    2048    256     2048    12.2771
fp32    2048    256     8192    8.0287
fp64    2048    256     8192    15.4497
fp32    2048    256     1120    -8.1388
fp64    2048    256     1120    5.8003
fp32    2048    256     1215    1.709
fp64    2048    256     1215    12.4369
fp32    2048    256     1856    5.1844
fp64    2048    256     1856    14.2236
fp32    2048    256     1302    2.8457
fp64    2048    256     1302    10.5728
fp32    2048    256     1329    -2.547
fp64    2048    256     1329    12.1123
fp32    2048    256     1531    2.4946
fp64    2048    256     1531    12.2398
fp32    2048    256     1313    6.1621
fp64    2048    256     1313    9.857
fp32    2048    256     1672    2.176
fp64    2048    256     1672    9.8899
fp32    2048    256     1851    4.6307
fp64    2048    256     1851    15.0223
fp32    2048    256     1584    3.5238
fp64    2048    256     1584    10.3181
fp32    2048    64      512     -11.5325
fp64    2048    64      512     8.5141
fp32    2048    64      2048    0.6066
fp64    2048    64      2048    25.8166
fp32    2048    64      8192    15.5994
fp64    2048    64      8192    29.453
fp32    2048    64      1120    1.5933
fp64    2048    64      1120    17.1686
fp32    2048    64      1215    -11.8064
fp64    2048    64      1215    21.7897
fp32    2048    64      1856    3.3061
fp64    2048    64      1856    17.6379
fp32    2048    64      1302    -1.201
fp64    2048    64      1302    26.775
fp32    2048    64      1329    -1.377
fp64    2048    64      1329    23.6142
fp32    2048    64      1531    0.9212
fp64    2048    64      1531    16.7177
fp32    2048    64      1313    2.8448
fp64    2048    64      1313    26.824
fp32    2048    64      1672    1.5334
fp64    2048    64      1672    23.7874
fp32    2048    64      1851    0.1934
fp64    2048    64      1851    25.1446
fp32    2048    64      1584    -2.8748
fp64    2048    64      1584    22.3902
fp32    8192    4096    512     0.0512
fp64    8192    4096    512     2.8049
fp32    8192    4096    2048    3.6683
fp64    8192    4096    2048    5.7372
fp32    8192    4096    8192    6.2501
fp64    8192    4096    8192    5.6644
fp32    8192    4096    1120    3.4347
fp64    8192    4096    1120    5.9099
fp32    8192    4096    1215    4.0591
fp64    8192    4096    1215    6.2049
fp32    8192    4096    1856    4.5046
fp64    8192    4096    1856    5.9
fp32    8192    4096    1302    3.8744
fp64    8192    4096    1302    5.74
fp32    8192    4096    1329    3.9169
fp64    8192    4096    1329    6.302
fp32    8192    4096    1531    5.0479
fp64    8192    4096    1531    6.048
fp32    8192    4096    1313    3.5261
fp64    8192    4096    1313    6.0544
fp32    8192    4096    1672    4.6081
fp64    8192    4096    1672    5.2568
fp32    8192    4096    1851    4.2022
fp64    8192    4096    1851    6.0934
fp32    8192    4096    1584    3.3852
fp64    8192    4096    1584    5.6772
fp32    8192    1024    512     3.7405
fp64    8192    1024    512     16.4627
fp32    8192    1024    2048    8.3918
fp64    8192    1024    2048    18.5254
fp32    8192    1024    8192    13.7773
fp64    8192    1024    8192    17.4314
fp32    8192    1024    1120    6.2023
fp64    8192    1024    1120    16.689
fp32    8192    1024    1215    9.5441
fp64    8192    1024    1215    19.7246
fp32    8192    1024    1856    9.864
fp64    8192    1024    1856    18.2895
fp32    8192    1024    1302    7.3145
fp64    8192    1024    1302    19.8528
fp32    8192    1024    1329    9.6131
fp64    8192    1024    1329    19.5526
fp32    8192    1024    1531    8.9847
fp64    8192    1024    1531    20.3696
fp32    8192    1024    1313    7.2819
fp64    8192    1024    1313    20.5361
fp32    8192    1024    1672    11.8095
fp64    8192    1024    1672    18.3047
fp32    8192    1024    1851    12.1042
fp64    8192    1024    1851    21.8124
fp32    8192    1024    1584    9.6549
fp64    8192    1024    1584    18.1818
fp32    8192    256     512     8.2649
fp64    8192    256     512     20.9372
fp32    8192    256     2048    15.6297
fp64    8192    256     2048    35.6407
fp32    8192    256     8192    21.7055
fp64    8192    256     8192    37.225
fp32    8192    256     1120    8.322
fp64    8192    256     1120    33.6497
fp32    8192    256     1215    12.9148
fp64    8192    256     1215    40.0554
fp32    8192    256     1856    12.2226
fp64    8192    256     1856    36.2642
fp32    8192    256     1302    12.2956
fp64    8192    256     1302    40.4711
fp32    8192    256     1329    10.2045
fp64    8192    256     1329    38.4891
fp32    8192    256     1531    14.9187
fp64    8192    256     1531    40.7874
fp32    8192    256     1313    9.5106
fp64    8192    256     1313    42.1367
fp32    8192    256     1672    15.2577
fp64    8192    256     1672    36.7527
fp32    8192    256     1851    15.668
fp64    8192    256     1851    40.2035
fp32    8192    256     1584    14.126
fp64    8192    256     1584    32.7602

tjingrant committed:
Separate and format test cases.

tjingrant commented:
Hi @zheng-xq , I think I resolved all your requested changes.

zheng-xq commented on tensorflow/python/kernel_tests/segment_reduction_ops_test.py:
```diff
@@ -646,56 +646,59 @@ def testGradientSegmentsInvalid4(self):
-      print("{:d}\t{:d}\t{:d}\t{:f}".format(
-          outer_dim, output_outer_dim, inner_dim, t[1]/t[0]))
+      print("__bench__ {} {:d}\t{:d}\t{:d}\t{:f}".format(
+          self._npTypeToStr(dtype), outer_dim, output_outer_dim, inner_dim, t[1]/t[0]))
```
Most benchmark comparison uses percentage in the difference. But it's up to you. Also mention "positive means the new implementation is faster".

Given some kernels do get slower. Add a TODO to investigate why that is the case. 

Thanks for your contribution!

zheng-xq commented:
LGTM, after minor changes.

tjingrant committed:
Add TODO for future investigations.

tjingrant commented:
I have modified the commit message to use percentage differences and a todo note is added to the file segment_reduction_ops.cc.

zheng-xq commented:
LGTM

tjingrant commented:
Hi @jhseu , any plan to move forward?

martinwicke commented:
Jenkins, test this please.

tjingrant commented:
Hi, the error message seems unrelated to my changes:
```
00:36:53 The following tests FAILED:
00:36:53 	252 - C:/tf_jenkins/home/workspace/tensorflow-pr-win-cmake-py/tensorflow/python/training/server_lib_test.py (Timeout)
00:36:53 Errors while running CTest
```
